## Promise
手写xxx方法论：
1. 该技术要解决什么问题？
2. 该技术是怎么解决它的？
3. 该技术（对比其他技术）有什么优缺点？
4. 如何解决这些缺点？

### 二、Promise要解决什么问题——回调地狱


### 三、Promise怎么解决回调地狱问题


### 四、Promise的优点
1. 减少缩进（将异步的写法变成同步的写法）
```javascript
// 函数作为参数层层嵌套
f1(xxx,function f2(a){
  f3(yyy,function f4(b){
      f5(a + b,function f6(){})
  })
})
```
使用promise的写法后：
```javascript
f1(xxx)
  .then(f2)  // 将作为参数的f2作为then的参数(满足条件后要执行的函数作为then的参数)  f3是作为f2的函数体，直接在f2里面调用了，这里无法看到  
  .then(f4)  // 将作为参数的f4作为then的参数(满足条件后要执行的函数作为then的参数)  f5是作为f4的函数体，直接在f5里面调用了，这里无法看到  
  .then(f6)  // 
```
使用then的写法，永远只有一个缩进，而使用回调的写法，随着嵌套的加深，缩进越来越多。

2. 消灭if(err)代码
* 错误处理单独放到一个函数中(then的第二个参数)
* 如果不处理，就一直往后抛
```javascript
f1(xxx)
.then(f2,error1)
.then(f4,error2)
.then(f6.error3)
.catch(errorAll)
```

### Promsie的特性
Promise是一个类
1. JS里类是特殊的函数
2. 类属性:length(可以忽略)   // length为1，只接受一个函数
3. 类方法:all/allSettled/race/reject/resolve    // Promise.all()、Promise.allSettled()这些都是类方法
4. 对象属性:then(重要)/finally/catch
5. 对象内部属性:state = pending/fulfilled/rejected

[Promise A+](https://promisesaplus.com/)
[Promise A+翻译](https://juejin.im/post/6844904070205931533#heading-6)




### Promise A+规范

- 2.2.4 在执行上下文堆栈（execution context）仅包含平台代码之前，不得调用 onFulfilled和onRejected。这句话的意思就是在我们的代码执行完毕之前不会执行then中的方法。示例：
```javascript
      const succeed = sinon.fake();
      const promise = new Promise((resolve) => {
        resolve();
      });
      promise.then(succeed);
      console.log('在这里没有执行完毕之前，不得调用then中的succeed方法');
```
上面的代码中：如果`console.log`没有执行完毕，那么then中的方法不会执行。这就是我们面试时经常碰到的promise问题，为什么promise的代码会在同步代码之后执行。那么这个是如何实现的了？我们都知道在resolve之后会马上调用then中的第一个方法，如何确保它是在同步代码之后执行。这说明resolve中一定存在异步代码，而且这个异步代码中通过改变promsie的状态来控制then函数的执行。其实现可能如下：
```javascript
  resolve = (result)=>{
        // resolve中存在异步代码
        setTimeout(() => {
            if(this.state !== 'pending'){
                return;
            }
            this.state = 'fulfilled';
            if(typeof this.succeed === 'function'){
                this.succeed(result);
            }
        },0)
  }
```

- 2.2.5 onFulfilled和onRejected必须作为函数调用（即没有this值）

这句话的意思是我们的then的两个函数调用时,this必须是undefined，不能指向其他的值。因此，我们在调用这两个函数时，需要手动指定this的值为undefined。
```javascript
  resolve = (result)=>{
        // then的第一个参数是在resolve执行的情况下执行
        setTimeout(() => {
            if(this.state !== 'pending'){
                return;
            }
            this.state = 'fulfilled';
            if(typeof this.succeed === 'function'){
                this.succeed.call(undefined,result);   // 手动指定this的值是undefined
            }
        },0)
  }
```
- 2.2.6 then方法可能在同一次promise中被调用多次
  - 2.2.6.1 如果/当promise完成时,所有相应的onFulfilled回调必须按照最原始的then顺序来执行
  - 2.2.6.2 如果/当promise失败时，所有相应的onRejected回调必须按照最原始的then顺序来执行
上面的意思是说，promise可以多次调用then方法，而且这些then方法必须按照调用的先后顺序分别执行对应的函数。示例：
```javascript
      const promise = new Promise(resolve => {
        resolve();
      });
      const callbacks = [sinon.fake(),sinon.fake(),sinon.fake()];
      promise.then(callbacks[0]);
      promise.then(callbacks[1]);
      promise.then(callbacks[2]);
```
上面的promise调用了三次then，对应的传入参数分别为callbacks[0],callbacks[2],callbacks[2]，那么resolve之后执行时，函数的执行顺序也必须是callbacks[0],callbacks[1],callback[2]。
这说明我们需要一个数组来保存then的参数，然后resolve之后遍历这个数组，依次执行对应的方法。
```javascript

class Promise{
  succeed = null;
  fail = null;
  state = 'pending';
  callbacks = [];   // 用来保存then中的函数
  resolve = (result)=>{
        // then的第一个参数是在resolve执行的情况下执行
        setTimeout(() => {
            if(this.state !== 'pending'){
                return;
            }
            this.state = 'fulfilled';
            // resolve之后遍历保存then参数的数组，然后依次调用。
            this.callbacks.forEach((handle) => {
                if(typeof handle[0] === 'function'){
                    handle[0].call(undefined,result);
                }
            })
        },0)
  }
  reject = (reason)=>{
        // then的第二个参数是在reject执行的情况下执行
        setTimeout(() => {
            if(this.state !== 'pending'){
                return;
            }
            this.state = "rejected";
            this.callbacks.forEach((handle) => {
                if(typeof handle[1] === 'function'){
                    handle[1].call(undefined,reason);
                }
            })
        },0)
  }
  constructor(fn){
      // new Promise()的参数必须是一个函数
    if(typeof fn !== 'function'){
        throw new Error('fn必须是一个函数');
    } 
    // new Promise(fn)这里的fn是立即执行的,且接收两个函数作为参数
    fn(this.resolve,this.reject);
  }
  then(succeed?,fail?){
    const handle = [null,null];
    if(typeof succeed === 'function'){
        handle[0] = succeed;
    }
    if(typeof fail === 'function'){
       handle[1] = fail;
    }
    this.callbacks.push(handle);
  }
}

export  default Promise
```

- 2.2.7 then方法必须返回一个promise
then方法必须返回一个promise,示例：
```javascript
promise2 = promise1.then(onFulfilled, onRejected); // promise2必须是一个新的promise
```